### DISCUSSION TITLE
Question on outputing directions of max and min principal stress

### DISCUSSION THREAD

**Discussion by Lingfu-Liu** [idaholab/moose#28579] [Q&A Modules: Solid mechanics]
I output max and min principal stress by using [AuxVariables] and [AuxKernels] ("Rank Two Aux") based on the example at https://mooseframework.inl.gov/source/auxkernels/RankTwoAux.html

```
> [AuxVariables]
>   [./maxprincipal]
>     order = CONSTANT
>     family = MONOMIAL
>   [../]
>   [./minprincipal]
>     order = CONSTANT
>     family = MONOMIAL
>   [../]
> [] # AuxVariables
> 
> [AuxKernels]
>   [./maxprincipal]
>     type = RankTwoScalarAux
>     rank_two_tensor = stress
>     variable = maxprincipal
>     scalar_type = MaxPrincipal
>   [../]
>   [./minprincipal]
>     type = RankTwoScalarAux
>     rank_two_tensor = stress
>     variable = minprincipal
>     scalar_type = MinPrincipal
>   [../]
> [] # AuxKernels
```

However, there is no example telling how the directions of principle stress are output. I found a website that seems to be the source code relevant to "Rank Two Aux" at https://mooseframework.inl.gov/docs/doxygen/moose/classRankTwoTensorTempl.html#ade890b615f07c8adb12b4dfa647f2fcd

In the source, Eigenvalues (principle stress) and Eigenvectors (direction of principle stress) are calcualted as by a function called "symmetricEigenvaluesEigenvectors()"

```
> void RankTwoTensorTempl< T >::symmetricEigenvaluesEigenvectors	
> (
> std::vector< T > & eigvals,
> RankTwoTensorTempl< T > & eigvecs 
> )		const
> computes eigenvalues and eigenvectors, assuming tens is symmetric, and places them in ascending order in eigvals.
> eigvecs is a matrix with the first column being the first eigenvector, the second column being the second, etc.
> Definition at line 866 of file RankTwoTensorImplementation.h.
>   868 {
>   869   mooseError(
>   870       "symmetricEigenvaluesEigenvectors is only available for ordered tensor component types");
>   871 }
```

I want to ask 
	1.	How can I find all the parameters that “RankTwoScalarAux” can calculate and output, along with their corresponding “name” in the syntax?
	2.	When I locate a function in the C++ source code, how can I relate it to the corresponding MOOSE syntax?

**Comment by GiudGiud** [comment]
Hello

The parameters are listed at the bottom of every object documentation page. Here:
https://mooseframework.inl.gov/source/auxkernels/RankTwoScalarAux.html

If you are looking for the source for that object, it's there. You have to look in auxkernels and the module that it is a part of
https://github.com/idaholab/moose/blob/next/modules/solid_mechanics/src/auxkernels/RankTwoScalarAux.C

The parameters are in the validParams static function of every class in MOOSE. The input file is parsed by the Parser and the Builder in moose (internal objects, which you can examine if you are comfortable with C++). 
The input syntax is fairly simple, it s usually:
```
params.addParam<type>("name_param", default_value, "documentation string");
```
this translates to:
```
[system of the object]
  [name for the object]
    type = <type of the object>
    name_param = some user selected value
```
pretty much the same way for every parameter

**Reply by Yulongggggg** [reply]
good question!
